File: frontend/web/src/api/card.ts
import { apiDelete, apiGet, apiPost, apiPut, type ApiEnvelope } from './client'
import type { Tag } from './tag'

export interface Card {
  id: number
  user_id: number
  content: string
  tags: string[]
  tag_ids?: number[]
  external_links: string[]
  internal_links: string[]
  created_at: string
  updated_at: string
}

export interface CreateCardRequest {
  content: string
  tags?: string[]
  tag_ids?: number[]
  external_links?: string[]
  internal_links?: string[]
}

export interface UpdateCardRequest {
  content?: string
  tags?: string[]
  tag_ids?: number[]
  external_links?: string[]
  internal_links?: string[]
}

export interface QueryCardRequest {
  content?: string
  tags?: string[]
  tag_ids?: number[]
  page?: number
  page_size?: number
}

export const createCard = (body: CreateCardRequest) => apiPost<ApiEnvelope<Card>>('/cards', body)

export const getCard = (id: number) => apiGet<ApiEnvelope<Card>>(`/cards/${id}`)

export const queryCards = (body: QueryCardRequest = {}) =>
  apiPost<ApiEnvelope<Card[]>>('/cards/query', body)

export const updateCard = (id: number, body: UpdateCardRequest) =>
  apiPut<ApiEnvelope<Card>>(`/cards/${id}`, body)

export const deleteCard = (id: number) => apiDelete<ApiEnvelope<null>>(`/cards/${id}`)

export interface CardTag {
  id: number
  user_id: number
  card_id: number
  tag_id: number
  created_at: string
  updated_at: string
}

export const listTagsByCard = (cardId: number) =>
  apiGet<ApiEnvelope<Tag[]>>(`/cards/${cardId}/tags`)

export const addTagToCard = (cardId: number, tagId: number) =>
  apiPost<ApiEnvelope<CardTag>>(`/cards/${cardId}/tags`, { tag_id: tagId })

export const replaceCardTags = (cardId: number, tagIds: number[]) =>
  apiPut<ApiEnvelope<CardTag[]>>(`/cards/${cardId}/tags`, { tag_ids: tagIds })

export const deleteTagFromCard = (cardId: number, tagId: number) =>
  apiDelete<ApiEnvelope<null>>(`/cards/${cardId}/tags/${tagId}`)

File: handler/dto/card.go
package dto

import "time"

type CreateCardRequest struct {
	Title         string   `json:"title"`
	Content       string   `json:"content" binding:"required"`
	Tags          []string `json:"tags"`
	TagIds        []int    `json:"tag_ids"`
	ExternalLinks []string `json:"external_links"`
	InternalLinks []string `json:"internal_links"`
}

type UpdateCardRequest struct {
	Title         *string   `json:"title"`
	Content       *string   `json:"content"`
	Tags          *[]string `json:"tags"`
	TagIds        *[]int    `json:"tag_ids"`
	ExternalLinks *[]string `json:"external_links"`
	InternalLinks *[]string `json:"internal_links"`
}

type ChangeCardContentVisibilityRequest struct {
	IsContentVisible *bool `json:"is_content_visible"`
}

type QueryCardDTO struct {
	Content  string   `json:"content"`
	Tags     []string `json:"tags"`
	TagIds   []int    `json:"tag_ids"`
	Page     int      `json:"page"`
	PageSize int      `json:"page_size"`
}

type CardResponse struct {
	Id               int       `json:"id"`
	UserId           int       `json:"user_id"`
	Title            string    `json:"title"`
	Content          string    `json:"content"`
	Tags             []string  `json:"tags"`
	TagIds           []int     `json:"tag_ids,omitempty"`
	ExternalLinks    []string  `json:"external_links"`
	InternalLinks    []string  `json:"internal_links"`
	IsContentVisible bool      `json:"is_content_visible"`
	CreatedAt        time.Time `json:"created_at"`
	UpdatedAt        time.Time `json:"updated_at"`
}

type CardQueryFilter struct {
	UserID   int
	Content  string
	Tags     []string
	TagIDs   []int
	Page     int
	PageSize int

	CreatedAt *time.Time
	UpdatedAt *time.Time
}
